### ISSUE TITLE
Message based cucumber-junit-formatter

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/common#871]
We need a `cucumber-junit-formatter` executable that reads messages from `STDIN` and writes JUnit XML to `STDOUT`. Any Cucumber implementation with a `message` formatter can then pipe generated messages to this formatter to generate JUnit XML.

Older Cucumbers with built-in JUnit formatters will have those removed when they support messages and people will have to use the `cucumber-junit-formatter` executable instead. The existing ones are:

* [JUnitFormatter.java](https://github.com/cucumber/cucumber-jvm/blob/master/core/src/main/java/io/cucumber/core/plugin/JUnitFormatter.java)
  * [test data](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/test/resources/io/cucumber/core/plugin)
* [junit.rb](https://github.com/cucumber/cucumber-ruby/blob/master/lib/cucumber/formatter/junit.rb)
  * [test data](https://github.com/cucumber/cucumber-ruby/blob/master/features/docs/formatters/junit_formatter.feature)

As pointed out in https://github.com/cucumber/cucumber-ruby/issues/1384 the generated XML should contain the steps (like the Java implementation does).

In order to be portable, `cucumber-junit-formatter` should be implemented in Go. Look at `demo-formatter/go` for inspiration.

**Comment by vincent-psarga** [comment]
I don't think it should be a stand alone tool. I think there should be a `cucumber/junit-formatter` in the mono-repo, written in all languages currently supported (Java, Ruby and JS). We could of course take advantage of the messages portability to have a common set of tests.
JUnit is one of the most widely used output format, and I definitely think it should built-in all cucumber releases, without having to add an extra tool

**Comment by mpkorstanje** [comment]
Might be of interest that the JUnit Team is currently working on a new format. I would wait too see how that turns out before we spend too much time implementing a structural solution. 

https://github.com/junit-team/junit5/issues/373

**Comment by mpkorstanje** [comment]
Also worth nothing that the JUnit formatter works just fine, even with the message protocol. So I don't see a need to remove it. I can see a use case for implementing it as a polyglot component with a common acceptance test suite.

**Comment by denim2x** [comment]
@aslakhellesoy Please explain why this tool should be standalone; wouldn't a transpiled multi-language (like [Haxe](https://haxe.org/manual/introduction-what-is-haxe.html)) module do?

**Comment by aslakhellesoy** [comment]
@vincent-psarga is [right](https://github.com/cucumber/cucumber/issues/871#issuecomment-577146088). It shouldn't be a standalone tool, but a library that can be used with all the supported Cucumber implementations.

Haxe is nice, but it lacks support for Ruby and Go, which are two of the languages we support. So I'm afraid we have to go with the one-implementation-per-language approach several other libraries in the monorepo uses.

**Comment by denim2x** [comment]
I'm considering having a Java [translator](https://github.com/jtransc/jtransc#what-is-this) which also [generates](https://github.com/scapix-com/scapix#scapix-language-bridge) multi-language bindings, integrated through JNI.

**Comment by aslakhellesoy** [comment]
That sounds interesting. If it turns out to work well it might reduce duplicate efforts quite a bit!

**Comment by denim2x** [comment]
@aslakhellesoy Could we continue this discussion on [Slack](https://join.slack.com/t/cyberdudegroup/shared_invite/zt-ib17mhuv-qHbPqgieh~TuY1I4r84Fyw)?

**Comment by denim2x** [comment]
@aslakhellesoy I've started a Kotlin implementation (a multi-platform plugin), developed progressively by porting original Java code while eliminating JVM dependencies (like `Class`). What would work [here](https://github.com/denim2x/cucumber-junit-formatter/blob/2a06ccbdb2f935c75597d18bb175d5fdd370c31e/src/commonMain/io/cucumber/plugin/core/JUnitFormatter.kt#L56) as an alternative - enums or strings?

**Comment by mpkorstanje** [comment]
@denim2x your formatter is using the events from Cucumber JVM. However Aslak is looking for an implementation that uses the messages format.

**Comment by mpkorstanje** [comment]
Also worth noting that while a Kotlin implementation would be executable on the JVM, we can not use it in Cucumber JVM. To avoid conflicts with the system under test Cucumber should pull in as few external dependencies as possible (i.e. zero). A Kotlin implementation would add a dependency on the Kotlin Standard Library.

**Comment by denim2x** [comment]
@mpkorstanje Thank you for letting me know; to this end, a new transpiler is being developed, targeting the JVM, C#, JS, Python, Ruby. This means that Kotlin is converted source-to-source into the target language/runtime, without relying on another standard library; the target's standard library is used.

**Comment by aslakhellesoy** [comment]
@denim2x where can we follow the development of this transpiler?

**Comment by denim2x** [comment]
@aslakhellesoy I'm working on a compiler [toolkit](https://github.com/denim2x/Kouncil/projects/1) for .NET, written in Kotlin.

**Comment by davidjgoss** [comment]
@aslakhellesoy I'm interested in working on this. Would it be okay implemented in JavaScript (using messages, not coupled to cucumber-js) in similar way to html-formatter?

**Comment by davidjgoss** [comment]
First pass at a JavaScript implementation here: https://github.com/davidjgoss/cucumberjs-junit-formatter

Here's an example from a test of what it outputs: https://github.com/davidjgoss/cucumberjs-junit-formatter/blob/main/test/snapshots/integration.test.ts.md

It's wrapped as a cucumber-js `Formatter` class, but doesn't use any of the old formatter helpers - just consumes envelopes and uses gherkin query and cucumber query.

We're now talking about having it live in the monorepo as a generic envelopes -> junit stream with a thin binding in cucumber-js (similar to the `html` formatter). We could then port other platforms' junit formatters here and share a test suite driven from the compatibility kit (which would ideally need a few more cases e.g. pending, undefined, ambiguous).

Things to maybe iron out:

- As @mpkorstanje points out, for each `<testcase>` the combination of `classname` and `name` should be unique. Currently these are populated with `pickle.uri` and `pickle.name` respectively, but the latter isn't guaranteed to be unique (e.g. with scenario outlines) so need to add some kind of qualifier.

**Comment by mpkorstanje** [comment]
The code that makes the names unique can be found here:

https://github.com/cucumber/cucumber-jvm/blob/main/core/src/main/java/io/cucumber/core/plugin/JUnitFormatter.java#L208

But probably best not to copy that as it assumes scenarios are executed in order. This is not guaranteed for messages.

If not unique, I would suggest numbering the nodes in a feature file and using the path to the pickle as a unique prefix.

```
1. Rule
  1.1 Scenario
     1.1.1 Examples Section
         1.1.1.1 Example 1
         1.1.1.2 Example 2
         1.1.1.3 Example 3
     1.1.2 Examples Section
         1.1.2.1 Example 1
         1.1.2.2 Example 2
         1.1.2.3 Example 3
2. Rule
  2.1 Scenario
 ....
```

**Comment by luke-hill** [comment]
@mpkorstanje if doing that wouldn't it be easier to use line number? As that's already in the AST right? Or am I getting confused?

**Comment by mpkorstanje** [comment]
Yes. Would be easier, but also less obvious I think.

**Comment by aurelien-reeves** [comment]
>     * As @mpkorstanje points out, for each `<testcase>` the combination of `classname` and `name` should be unique. Currently these are populated with `pickle.uri` and `pickle.name` respectively, but the latter isn't guaranteed to be unique (e.g. with scenario outlines) so need to add some kind of qualifier.

If it is the combination of `classname` and `name` which should be unique, then `pickle.uri` + `pickle.name` should work. Am I missing something?

**Comment by davidjgoss** [comment]
@aurelien-reeves pickles in the same feature file could have the same name if:

- You have a scenario outline with examples (unless the name includes a parameter)
- People have just given several scenarios exactly the same name (bad, not invalid Gherkin)

**Comment by aurelien-reeves** [comment]
Oh, ok.

What would you think of adding the line number in the URI, as an anchor?

**Comment by mpkorstanje** [comment]
Okay. Other suggestion then. We can do what ever JUnit does when it encounters duplicate. 

I've seen JUnit do something like this (I think, would have to double check)

```
ExampleTest
  ParameterizedTest [0]
  ParameterizedTest [1]
  ParameterizedTest [2]
  ParameterizedTest [3]
 ```

**Comment by luke-hill** [comment]
> @aurelien-reeves pickles in the same feature file could have the same name if:
> 
> * You have a scenario outline with examples (unless the name includes a parameter)
> * People have just given several scenarios exactly the same name (bad, not invalid Gherkin)

I have the following gherkin frequently (Which I don't consider bad)

```
Feature: Foo
  Rule: English language
    Scenario: Can do Foo

  Rule: Welsh language
    Scenario: Can do Foo
```

**Comment by davidjgoss** [comment]
@luke-hill true - I forget that the rule name doesn't contribute to the pickle name

**Comment by aurelien-reeves** [comment]
> 
> 
> Okay. Other suggestion then. We can do what ever JUnit does when it encounters duplicate.
> 
> I've seen JUnit do something like this (I think, would have to double check)
> 
> ```
> ExampleTest
>   ParameterizedTest [0]
>   ParameterizedTest [1]
>   ParameterizedTest [2]
>   ParameterizedTest [3]
> ```

Only for duplicates right?

**Comment by davidjgoss** [comment]
@mpkorstanje @aurelien-reeves thoughts on omitting the ` [0]` for the first one? Would make it a bit simpler to implement I think.

**Comment by davidjgoss** [comment]
> Only for duplicates right?

Yep! Here's a PR that does that https://github.com/davidjgoss/cucumberjs-junit-formatter/pull/1

**Comment by mpkorstanje** [comment]
Nasty case:

```
Scenario: A
Scenario: A
Scenario: A[1]
```

**Comment by davidjgoss** [comment]
@mpkorstanje oh come on, that's just gratuitous! 😆

**Comment by davidjgoss** [comment]
We have a PR from the community to add a JUnit formatter to cucumber-js:
https://github.com/cucumber/cucumber-js/pull/2121

It doesn't consume pure messages yet but I think it's a better implementation than my POC from last year in other ways. I think it would be worth accepting it as a step towards this goal - we can spin it out into the common repo when we're ready. In terms of the output (e.g. how it expresses failures, what it puts in `<system-out>`) it seems like it should follow what cucumber-jvm does in the absence of any other standard.

Any thoughts?

**Comment by mpkorstanje** [comment]
I agree. Note that the Cucumber JVM implementation is tricky and the JUnit XML a very poorly defined standard. I'll add my remarks to the other PR.

**Comment by mpkorstanje** [comment]
I've found documentation for the JUnit XML format:

https://llg.cubic.org/docs/junit/

**Comment by mpkorstanje** [comment]
Java is done. Now to extract the formatter added in https://github.com/cucumber/cucumber-js/pull/2121

**Comment by mpkorstanje** [comment]
JUnit was done with https://github.com/cucumber/cucumber-js/pull/2445
